---
title: "Transcripts Per Gene"
author: "Mervin M Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Purpose

Summarize transcripts per gene for various UTRome parameter configurations.

# Initialization

## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(magrittr)
library(tidyverse)
library(cowplot)
```

## Parameters
```{r set_params}
set.seed(20220404)
```

## Functions

```{r methods}
read_merges <- function (epsilon, tpm) {
    ipa_file <- sprintf("data/gff/utrome.e%s.t%s.gc25.pas3.f0.9999.w500.ipa.tsv",
                        epsilon, tpm)
    merge_file <- sprintf("data/gff/utrome.e%s.t%s.gc25.pas3.f0.9999.w500.m200.tsv",
                          epsilon, tpm)
    txs_ipa <- read_tsv(ipa_file)$transcript_id
    read_tsv(merge_file) %>%
        mutate(ipa_in=tx_in %in% txs_ipa,
               ipa_out=tx_out %in% txs_ipa)
}

count_txs <- function (df, use_unmerged=FALSE) {
    idx_cols <- if (use_unmerged) c("tx_in") else c("tx_out")
    distinct(df, across(all_of(idx_cols))) %>% nrow()
}

count_genes <- . %>% { nrow(distinct(., gene_out)) } 

count_ipa_txs <- function (df, use_unmerged=FALSE) {
    if (use_unmerged) {
        df %>% 
            filter(ipa_in) %>%
            distinct(tx_in) %>%
            nrow()
    } else {
        df %>% 
            filter(ipa_out) %>%
            distinct(tx_out) %>%
            nrow()
    }
}

count_ipa_genes <- function (df, use_unmerged=FALSE) {
    if (use_unmerged) {
        df %>%
            group_by(gene_out) %>%
            summarize(is_ipa=any(ipa_in), .groups='drop') %$%
            sum(is_ipa)
    } else {
        df %>%
            group_by(gene_out) %>%
            summarize(is_ipa=any(ipa_out), .groups='drop') %$%
            sum(is_ipa)
    }
} 

count_txs_per_gene <- function (df, omit_ipa=TRUE, use_unmerged=FALSE) {
    if (use_unmerged) {
        df %>%
            { if (omit_ipa) { filter(., ipa_in) } else . } %>%
            distinct(tx_in, gene_out) %>%
            count(gene_out, name="n_txs") %>%
            count(n_txs, name="n_genes") %T>%
            mutate(pct_genes=n_genes/sum(n_genes))
    } else {
        df %>%
            { if (omit_ipa) { filter(., ipa_out) } else . } %>%
            distinct(tx_out, gene_out) %>%
            count(gene_out, name="n_txs") %>%
            count(n_txs, name="n_genes") %T>%
            mutate(pct_genes=n_genes/sum(n_genes))
    }
}
```

# Data
## Loading
```{r load_data, message=FALSE}
df_merges <- tibble(epsilon=rep(c(30), 1),
       tpm=rep(c(5), each=1)) %>%
    mutate(merge_data=map2(epsilon, tpm, read_merges))
```

## Preprocessing
```{r prepare_data}
df_merges %>%
    mutate(n_txs_unmerged=map_int(merge_data, count_txs, use_unmerged=TRUE),
           n_txs_merged=map_int(merge_data, count_txs, use_unmerged=FALSE),
           n_ipa_txs_unmerged=map_int(merge_data, count_ipa_txs, use_unmerged=TRUE),
           n_ipa_txs_merged=map_int(merge_data, count_ipa_txs, use_unmerged=FALSE),
           n_genes=map_int(merge_data, count_genes),
           n_ipa_genes_unmerged=map_int(merge_data, count_ipa_genes, use_unmerged=TRUE),
           n_ipa_genes_merged=map_int(merge_data, count_ipa_genes, use_unmerged=FALSE)
           ) %>%
    select(epsilon, tpm, n_txs_unmerged, n_txs_merged, 
           n_ipa_txs_unmerged, n_ipa_txs_merged,
           n_genes, n_ipa_genes_unmerged, n_ipa_genes_merged) %>%
    knitr::kable()
```

# Analysis

### Total Transcripts
```{r}
EPS=30
TPM=5

n_txs <- df_merges %>%
    filter(epsilon == EPS, tpm == TPM) %$%
    merge_data[[1]] %>%
    distinct(tx_out, gene_out, ipa_out) %>%
    nrow()

sprintf("%d total transcripts", n_txs)
```

### IPA Transcripts
```{r}
EPS=30
TPM=5

n_ipa_txs <- df_merges %>%
    filter(epsilon == EPS, tpm == TPM) %$%
    merge_data[[1]] %>%
    distinct(tx_out, gene_out, ipa_out) %$%
    sum(ipa_out)

sprintf("%d IPA transcripts", n_ipa_txs)
sprintf("%0.2f%% IPA transcripts", 100*n_ipa_txs/n_txs)
```

### Multi-UTR - with IPA
```{r}
EPS=30
TPM=5
df_merges %>%
    filter(epsilon == EPS, tpm == TPM) %$%
    merge_data[[1]] %>%
    distinct(tx_out, gene_out, ipa_out) %>%
    count(gene_out, name="n_txs") %>%
    count(n_txs, name="n_genes") %T>%
    { mutate(., pct=n_genes/sum(n_genes)) %>% print() } %>%
    mutate(n_txs_label=ifelse(n_txs >= 9, "9+", as.character(n_txs))) %>%
    ggplot(aes(x=n_txs_label, y=n_genes)) +
    stat_identity(geom='bar', fill='lightgrey', color='black') +
    #scale_x_continuous(expand=c(0,0.5)) +
    scale_y_continuous(expand=c(0,0,0.05,0)) +
    labs(x="Number of Isoforms (with IPA)", y="Genes") +
    theme_bw()
```

### Multi-UTR - without IPA
```{r}
EPS=30
TPM=5
df_merges %>%
    filter(epsilon == EPS, tpm == TPM) %$%
    merge_data[[1]] %>%
    distinct(tx_out, gene_out, ipa_out) %>%
    filter(!ipa_out) %>%
    count(gene_out) %>%
    count(n) %T>%
    { mutate(., pct=nn/sum(nn)) %>% print() } %>%
    mutate(n_txs_label=ifelse(n >= 9, "9+", as.character(n))) %>%
    ggplot(aes(x=n_txs_label, y=nn)) +
    stat_identity(geom='bar', fill='lightgrey', color='black') +
    #scale_x_continuous(expand=c(0,0.5)) +
    scale_y_continuous(expand=c(0,0,0.05,0)) +
    labs(x="Number of Isoforms (No IPA)", y="Genes") +
    theme_bw()
```

# Conclusion

<!--
What was found?
Clearly state conclusions.
Was there anything unexpected?
Are there unanswered questions?
What should be worked on next?
-->

---

# Runtime Details
## Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```

## Conda Environment
```{bash comment="", echo=FALSE}
if ! command -v conda &> /dev/null
then
  echo "Conda not detected."
elif [ -z "${CONDA_PREFIX}" ]
then
  echo "No active Conda environment."
else
  echo "## Conda Environment YAML"
  conda env export
fi
```
