---
title: "Cleavage Site Counts"
author: "Mervin M Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    code_folding: show
    toc: true
    toc_float: true
---

# Initialization
## Libraries
```{r libs, message=FALSE, warning=FALSE}
library(plyranges)
library(tidyverse)
library(magrittr)
library(gtsummary)
library(GenomicRanges)
```

## Parameters
```{r params}
EPSILON = 30
TPM = 5

FILE_UTROME = sprintf("data/granges/utrome_gr_txs.e%d.t%d.gc25.pas3.f0.9999.w500.Rds",
                      EPSILON, TPM)
FILE_BED = sprintf("data/bed/celltypes/celltypes.e%d.t%d.bed.gz", EPSILON, TPM)
```

# Load Data
## Cleavage Sites by Cell Type
```{r load_sites}
gr_sites <- read_bed(FILE_BED) %>% 
    `seqlevelsStyle<-`("UCSC") %>%
    keepStandardChromosomes(pruning.mode="coarse") %>%
    anchor_center() %>%
    mutate(width=EPSILON)
```

## Mouse UTRome Transcripts
```{r load_txs, message=FALSE, warning=FALSE}
## Load all transcripts
gr_txs <- readRDS(FILE_UTROME)

## focus on cleavage sites
df_txs <- gr_txs %>%
    anchor_3p %>%
    mutate(n_celltypes=count_overlaps_directed(mutate(., width=0), gr_sites),
           origin=case_when(
               is_novel ~ "UTRome",
               !is_novel & (n_celltypes > 0) ~ "Common",
               TRUE ~ "GENCODE") %>%
               fct(levels=c("Common", "UTRome", "GENCODE"))
           ) %>%
    as_data_frame() %>%
    dplyr::select(gene_id, transcript_id, utr_rank,
                  is_ipa, is_novel, origin) %>%
    group_by(gene_id) %>%
    mutate(min_utr_rank=ifelse(all(is_ipa), 0, min(utr_rank[!is_ipa])),
           max_utr_rank=ifelse(all(is_ipa | is_novel), Inf, max(utr_rank[!is_ipa & !is_novel])),
           is_single=!is_ipa & !is_novel & (sum(!is_ipa) == 1),
           is_distal=!is_single & (utr_rank == max_utr_rank),
           is_proximal=!is_single & !is_distal & (utr_rank == min_utr_rank),
           is_medial=!is_ipa & (utr_rank > min_utr_rank) & (utr_rank < max_utr_rank),
           is_extended=!is_ipa & is_novel & (utr_rank > max_utr_rank)) %>%
    ungroup() %>%
    mutate(is_exclusive=is_ipa + is_single + is_proximal + is_medial + is_distal + is_extended == 1) %T>%
    { stopifnot(all(.$is_exclusive)) } %>%
    mutate(utr_class=case_when(
        is_single ~ "Single",
        is_ipa ~ "IU",
        is_proximal ~ "SU",
        is_medial ~ "MU",
        is_distal ~ "LU",
        is_extended ~ "ELU"
        ) %>% fct(levels=c("Single", "IU", "SU", "MU", "LU", "ELU"))) %>%
    mutate(novel_lbl=ifelse(is_novel, "Novel", "GENCODE"))
```

## UTR Type
```{r plot_type, fig.width=3, fig.height=4}
df_txs %>%
    ggplot(aes(x=utr_class)) +
    geom_bar(fill='grey80', color='black', linewidth=0.2, width=0.85) +
    scale_y_continuous(expand=c(0,0,0.05,0)) +
    labs(x=NULL, y="Cleavage Sites") +
    theme_bw()

ggsave("img/sq/fig1b-cleavage-sites-mca.pdf", width=3, height=4, dpi=300)

tbl_summary(df_txs, include=utr_class, label=list(utr_class~"UTR Type"))
```



```{r plot_type_novel, fig.width=4.5, fig.height=4}
df_txs %>%
    ggplot(aes(x=utr_class, fill=fct_rev(novel_lbl))) +
    geom_bar(color='black', linewidth=0.2, width=0.85) +
    scale_y_continuous(expand=c(0,0,0.05,0)) +
    scale_fill_grey(start=0.7, end=0.4) +
    labs(x=NULL, y="Cleavage Sites", fill="Origin") +
    theme_bw()

ggsave("img/sq/sup1b-cleavage-sites-mca-2.pdf", width=4.5, height=4, dpi=300)

df_txs %>%
    tbl_summary(include=utr_class, by=novel_lbl, label=list(utr_class~"UTR Type"))
```

```{r plot_type_origin, fig.width=4.5, fig.height=4}
df_txs %>%
    ggplot(aes(x=utr_class, fill=fct_rev(origin))) +
    geom_bar(color='black', linewidth=0.2, width=0.85) +
    scale_y_continuous(expand=c(0,0,0.05,0)) +
    scale_fill_grey(start=0.7, end=0.4) +
    labs(x=NULL, y="Cleavage Sites", fill="Origin") +
    theme_bw()

ggsave("img/sq/sup1b-cleavage-sites-mca-3.pdf", width=4.5, height=4, dpi=300)

tbl_summary(df_txs, include=utr_class, by=origin, label=list(utr_class~"UTR Type"))
```

---

# Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```
